### PR TITLE
teamviewer: fix #96633, #44307 and #97148 + 15.15.5 -> 15.18.5 -> 15.22.3

### LINKED ISSUE #96633: Teamviewer never connects to it's servers
With current unstable: `nix-shell -I nixpkgs=channel:nixos-unstable -p teamviewer --pure`
Running reamviewer will leave it stuck complaining about needing to check your connection, or in the "connecting .." state.
I've had this issue for a while now and I'm only now getting around to filing it. I'm stumped. Does teamviewer work for anyone else?
image

You can see the stdout in the image. Also the GUI doesn't show up unless I use `--pure`.

### PR DESCRIPTION
**Pull request by jraygauthier**
###### Motivation for this change

 -  Fixing issue #96633
 -  Fixing issue #44307
 -  Fixing observed service stop issue
 -  Updating the package to latest possible version backportable to 21.05 (bound by nixpkgs qt5.14) 
 -  Some refactors suggested by @Artturin
 -  Fixing issue #97148 as suggested by @Artturin 
 -  For unstable: Updated to latest which requires qt5.15 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Tested the service with these 2 patches applied on top of my current 21.05 config. Works fine.

Tested client server launching these manually from package built from this branch. Works fine and fix the bug.

Note that `nixpkgs-review` raise an error which I am completely unable to understand:

```bash
$ nix-shell -I nixpkgs=. -p nixpkgs-review --run "nixpkgs-review rev HEAD^"
# ,,
1 package updated:
teamviewer
# ..
builder for '/nix/store/r86mqdfpqpi87kq3zb4d5fdzpa22r33x-review-shell.drv' failed with exit code 1; last 2 log lines:
  qtPreHook
  Error: wrapQtAppsHook is not used, and dontWrapQtApps is not set.
error: build of '/nix/store/r86mqdfpqpi87kq3zb4d5fdzpa22r33x-review-shell.drv' failed
1 package built:
teamviewer
# ..
```

If you look at the package, it effectively sets `dontWrapQtApps = true;`.

### CODE REVIEW

**Comment by Artturin** [comment]
qt5 is currently at `5.15.2` at nixos-unstable

File: pkgs/applications/networking/remote/teamviewer/default.nix [hunk 1]
```diff
@@ -52,23 +73,26 @@ mkDerivation rec {
     substituteInPlace $out/share/teamviewer/tv_bin/script/tvw_config \
       --replace '/var/run/' '/run/'
 
-    wrapProgram $out/share/teamviewer/tv_bin/script/teamviewer --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [ libXrandr libX11 ]}"
-    wrapProgram $out/share/teamviewer/tv_bin/teamviewerd --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [ libXrandr libX11 ]}"
-    wrapProgram $out/share/teamviewer/tv_bin/TeamViewer --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [ libXrandr libX11 ]}"
-    wrapProgram $out/share/teamviewer/tv_bin/TeamViewer_Desktop --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [libXrandr libX11 libXext libXdamage libXtst libSM libXfixes ]}"
+    wrapProgram $out/share/teamviewer/tv_bin/script/teamviewer --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [ libXrandr libX11 dbus ]}"
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/remote/teamviewer/default.nix:76]:
```suggestion
    wrapProgram $out/share/teamviewer/tv_bin/script/teamviewer \
      --prefix LD_LIBRARY_PATH : "${lib.makeLibraryPath [ libXrandr libX11 dbus ]}"
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please follow the contributing guide when naming your commits/PRs.

File: pkgs/applications/networking/remote/teamviewer/default.nix [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
 
 mkDerivation rec {
   pname = "teamviewer";
-  version = "15.15.5";
+  version = "15.18.5";
 
   src = fetchurl {
     url = "https://dl.tvcdn.de/download/linux/version_15x/teamviewer_${version}_amd64.deb";
-    sha256 = "sha256-H/CSc2RcjI+Fm8awYcXm3ioAJpbSNEMwGVrTozMux3A=";
+    sha256 = "sha256:03m8k6qr3mgylik1gqi8qlm0dngb3vf4p6gdwd5hyvz7gh5mdm2r";
```
[the remark this answers is not in the document]
Comment (Artturin) [reply pkgs/applications/networking/remote/teamviewer/default.nix:13]: Why?

**Comment by jraygauthier** [comment]
> this patch reduces duplication and add coreutils to fix #97148

Done in c55bc5bfd3377d54f5b6153d09c033cceadfcc05 and 4fb188e1d190f2e08053dba3e2e95fff1e29183a.

Retested the whole thing with these 2 new patches after clearing `/var/lib/teamviewer`. Everything still work fine.

**Comment by jraygauthier** [comment]
@SuperSandro2000 / @Artturin: Will be working on <https://github.com/jraygauthier/nixpkgs/commits/jrg/post-140076-pr-teamviewer> for change request that did not make the merge and create a separate PR once I have some interesting module test ready.

**Pull request merged**